src/mpid/ch4/src/mpidig_part.h

```diff
@@ -9,54 +9,38 @@
 #include "mpidig_part_utils.h"
 
-void MPIDIG_precv_matched(MPIR_Request * part_req);
 int MPIDIG_mpi_psend_init(const void *buf, int partitions, MPI_Aint count,
                           MPI_Datatype datatype, int dest, int tag,
                           MPIR_Comm * comm, MPIR_Info * info, MPIR_Request ** request);
 int MPIDIG_mpi_precv_init(void *buf, int partitions, MPI_Aint count,
                           MPI_Datatype datatype, int source, int tag,
                           MPIR_Comm * comm, MPIR_Info * info, MPIR_Request ** request);
 
-MPL_STATIC_INLINE_PREFIX int MPIDIG_part_start(MPIR_Request * request)
-{
-    int mpi_errno = MPI_SUCCESS;
-    MPIR_FUNC_ENTER;
-
-    MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
-
-    /* Indicate data transfer starts.
-     * Decrease when am request completes on sender (via completion_notification),
-     * or received data transfer AM on receiver. */
-    MPIR_cc_set(request->cc_ptr, 1);
-
-    /* No need to increase refcnt for comm and datatype objects,
-     * because it is erroneous to free an active partitioned req if it is not complete.*/
-
-    if (request->kind == MPIR_REQUEST_KIND__PART_RECV && MPIDIG_PART_REQUEST(request, peer_req_ptr)) {
-        mpi_errno = MPIDIG_part_issue_cts(request);
-    }
-
-    MPIR_Part_request_activate(request);
-
-    MPID_THREAD_CS_EXIT(VCI, MPIDI_VCI(0).lock);
-    MPIR_FUNC_EXIT;
-    return mpi_errno;
-}
+int MPIDIG_part_start(MPIR_Request * request);
 
-MPL_STATIC_INLINE_PREFIX int MPIDIG_mpi_pready_range(int partition_low, int partition_high,
+MPL_STATIC_INLINE_PREFIX int MPIDIG_mpi_pready_range(int p_low, int p_high,
                                                      MPIR_Request * part_sreq)
 {
     int mpi_errno = MPI_SUCCESS;
     MPIR_FUNC_ENTER;
-
-    int incomplete = 1, i;
-    for (i = partition_low; i <= partition_high; i++)
-        MPIR_cc_decr(&MPIDIG_PART_REQUEST(part_sreq, u.send).ready_cntr, &incomplete);
-
-    if (!incomplete) {
-        MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
-        mpi_errno = MPIDIG_part_issue_data(part_sreq, MPIDIG_PART_REGULAR);
-        MPID_THREAD_CS_EXIT(VCI, MPIDI_VCI(0).lock);
+    MPIR_Assert(MPIR_Part_request_is_active(part_sreq));
+    MPIR_Assert(part_sreq->kind == MPIR_REQUEST_KIND__PART_SEND);
+
+    const bool is_matched = MPIDIG_PART_REQUEST(part_sreq, peer_req_ptr);
+    MPIR_cc_t *cc_part = MPIDIG_PART_REQUEST(part_sreq, u.send.cc_part);
+
+    /* for each partition mark it as ready and start them if we can */
+    for (int i = p_low; i <= p_high; ++i) {
+        int incomplete;
+        MPIR_cc_decr(&cc_part[i], &incomplete);
+
+        // send the partition if matched and is complete
+        if (is_matched && !incomplete) {
+            // TODO change the VCI here
+            MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
+            mpi_errno &= MPIDIG_part_issue_data(i, part_sreq, MPIDIG_PART_REGULAR);
+            MPID_THREAD_CS_EXIT(VCI, MPIDI_VCI(0).lock);
```

Comment:
`mpi_errno` is initialized to `0`, so `&=` here is never going to change it. I'd suggest just checking the return code after each issue call. If an error is detected, it'll jump out of the loop and return the error to the user.
```suggestion
            MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
            mpi_errno = MPIDIG_part_issue_data(i, part_sreq, MPIDIG_PART_REGULAR);
            MPID_THREAD_CS_EXIT(VCI, MPIDI_VCI(0).lock);
            MPIR_ERR_CHECK(mpi_errno);
```

Reply:
solved in 87ccea181bd6cf17abb739abcd610e17782195f4